### PR TITLE
Add FusionDefinition.multidevice_schedule

### LINKED ISSUE #2929: mpi_test.py stopped working.
To reproduce, 

```
$ mpirun -np 2 pytest tests/python/mpi_test.py

_____________________________________________________________________________________ ERROR collecting tests/python/mpi_test.py ______________________________________________________________________________________
tests/python/mpi_test.py:43: in <module>
    o = fn.execute(inputs)
nvfuser/__init__.py:215: in execute
    self._setup_schedule(inputs)
E   RuntimeError:  INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/fusion_executor/executor_utils.cpp":731, please report a bug with repro script to NVFuser at https://github.com/NVIDIA/Fuser/issues. When trying to run the provided host program, there was an error with the provided input 0. Provided input was:
E     Tensor(sizes=[1, 4], stride=[4, 1], dtype=float, device=cuda:0, data_ptr=0x7d88c0e00000)
E     Fusion input is:
E     T0_g_float[ iS0{2}, iS1{4} ]
E     Expr eval provided the error:
E   """Tried to bind to a constant value: 2 as 1"""
E
E   Exception raised from bindInputs at /opt/pytorch/nvfuser/csrc/fusion_executor/executor_utils.cpp:731 (most recent call first):
E   frame #0: nvfuser::nvfCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xf3 (0x7d88d979b17f in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
E   frame #1: nvfuser::nvfErrorFail(char const*, char const*, unsigned int, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x53 (0x7d88d9b00223 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
E   frame #2: <unknown function> + 0x13ce26 (0x7d88d97a6e26 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
E   frame #3: <unknown function> + 0x45668c (0x7d88d9ac068c in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
E   frame #4: nvfuser::python_frontend::FusionDefinition::setupSchedule(c10::ArrayRef<c10::IValue> const&) + 0x186 (0x7d88d9f3ea36 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
E   frame #5: <unknown function> + 0x1b367d (0x7d88d981d67d in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
E   frame #6: <unknown function> + 0x22e603 (0x7d88d9898603 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
E   frame #7: <unknown function> + 0x2bb660 (0x7d88d9925660 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
E   frame #8: <unknown function> + 0x15adae (0x64c764b7bdae in /usr/bin/python3)
E   frame #9: _PyObject_MakeTpCall + 0x25b (0x64c764b7252b in /usr/bin/python3)
E   frame #10: <unknown function> + 0x16952b (0x64c764b8a52b in /usr/bin/python3)
E   frame #11: _PyEval_EvalFrameDefault + 0x64e2 (0x64c764b6a742 in /usr/bin/python3)
E   frame #12: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #13: _PyEval_EvalFrameDefault + 0x8cb (0x64c764b64b2b in /usr/bin/python3)
E   frame #14: <unknown function> + 0x140096 (0x64c764b61096 in /usr/bin/python3)
E   frame #15: PyEval_EvalCode + 0x86 (0x64c764c56f66 in /usr/bin/python3)
E   frame #16: <unknown function> + 0x23ba2d (0x64c764c5ca2d in /usr/bin/python3)
E   frame #17: <unknown function> + 0x15b909 (0x64c764b7c909 in /usr/bin/python3)
E   frame #18: _PyEval_EvalFrameDefault + 0x6d5 (0x64c764b64935 in /usr/bin/python3)
E   frame #19: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #20: _PyEval_EvalFrameDefault + 0x8cb (0x64c764b64b2b in /usr/bin/python3)
E   frame #21: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #22: _PyEval_EvalFrameDefault + 0x6d5 (0x64c764b64935 in /usr/bin/python3)
E   frame #23: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #24: _PyEval_EvalFrameDefault + 0x6d5 (0x64c764b64935 in /usr/bin/python3)
E   frame #25: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #26: _PyEval_EvalFrameDefault + 0x6d5 (0x64c764b64935 in /usr/bin/python3)
E   frame #27: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #28: _PyEval_EvalFrameDefault + 0x64e2 (0x64c764b6a742 in /usr/bin/python3)
E   frame #29: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #30: _PyEval_EvalFrameDefault + 0x64e2 (0x64c764b6a742 in /usr/bin/python3)
E   frame #31: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #32: _PyEval_EvalFrameDefault + 0x19b6 (0x64c764b65c16 in /usr/bin/python3)
E   frame #33: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #34: _PyEval_EvalFrameDefault + 0x6d5 (0x64c764b64935 in /usr/bin/python3)
E   frame #35: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #36: _PyEval_EvalFrameDefault + 0x8cb (0x64c764b64b2b in /usr/bin/python3)
E   frame #37: <unknown function> + 0x166d4e (0x64c764b87d4e in /usr/bin/python3)
E   frame #38: _PyObject_GenericGetAttrWithDict + 0x501 (0x64c764b7a4a1 in /usr/bin/python3)
E   frame #39: PyObject_GetAttr + 0x4d (0x64c764b7895d in /usr/bin/python3)
E   frame #40: _PyEval_EvalFrameDefault + 0x61c5 (0x64c764b6a425 in /usr/bin/python3)
E   frame #41: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #42: _PyEval_EvalFrameDefault + 0x8cb (0x64c764b64b2b in /usr/bin/python3)
E   frame #43: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #44: _PyEval_EvalFrameDefault + 0x8cb (0x64c764b64b2b in /usr/bin/python3)
E   frame #45: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #46: _PyEval_EvalFrameDefault + 0x6d5 (0x64c764b64935 in /usr/bin/python3)
E   frame #47: <unknown function> + 0x169251 (0x64c764b8a251 in /usr/bin/python3)
E   frame #48: _PyEval_EvalFrameDefault + 0x64e2 (0x64c764b6a742 in /usr/bin/python3)
E   frame #49: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #50: _PyEval_EvalFrameDefault + 0x2a49 (0x64c764b66ca9 in /usr/bin/python3)
E   frame #51: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #52: _PyEval_EvalFrameDefault + 0x64e2 (0x64c764b6a742 in /usr/bin/python3)
E   frame #53: <unknown function> + 0x169251 (0x64c764b8a251 in /usr/bin/python3)
E   frame #54: _PyEval_EvalFrameDefault + 0x64e2 (0x64c764b6a742 in /usr/bin/python3)
E   frame #55: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #56: _PyObject_FastCallDictTstate + 0x16d (0x64c764b7176d in /usr/bin/python3)
E   frame #57: _PyObject_Call_Prepend + 0x5c (0x64c764b875ac in /usr/bin/python3)
E   frame #58: <unknown function> + 0x281370 (0x64c764ca2370 in /usr/bin/python3)
E   frame #59: _PyObject_MakeTpCall + 0x25b (0x64c764b7252b in /usr/bin/python3)
E   frame #60: _PyEval_EvalFrameDefault + 0x7611 (0x64c764b6b871 in /usr/bin/python3)
E   frame #61: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #62: _PyEval_EvalFrameDefault + 0x6d5 (0x64c764b64935 in /usr/bin/python3)
E   frame #63: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
============================================================================================== short test summary info ===============================================================================================
ERROR tests/python/mpi_test.py - RuntimeError:  INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/fusion_executor/executor_utils.cpp":731, please report a bug with repro script to NVFuser at https://github.com/NVIDIA/Fuser/issues. When try...
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
================================================================================================== 1 error in 1.31s ==================================================================================================
```

I guess it started with https://github.com/NVIDIA/Fuser/pull/2635. However, I believe that's a right fix and shouldn't be rolled back. 

Instead, the root cause I believe is that bindInputs (called by concretization, which is in turn called by `FusionDefinition._setup_schedule`) happens before `FusionDefinition.schedule`: https://github.com/NVIDIA/Fuser/blob/7090a639ca9aa5154ef11c3cdf34374a1a4fc357/nvfuser/__init__.py#L215-L216. Without knowing the input is parallelized on DIDx, concretization incorrectly attempts to bind the device dimension extent with 1.

### PR DESCRIPTION
**Pull request by wujingyue**
and use that to specify DID schedules.

Fixes #2929.

### CODE REVIEW

**Comment by wujingyue** [comment]
!build

File: nvfuser/__init__.py [hunk 1]
```diff
@@ -202,6 +213,8 @@ def execute(
             self.definition()
             self._finalize_definition()
 
+        self.multidevice_schedule()
```
Comment (rdspring1) [inline nvfuser/__init__.py:216]: I think `multidevice_schedule` and `schedule` conflict with each other. Should there be an XOR check here?

Comment (rdspring1) [reply nvfuser/__init__.py:216]:
```python
# NOTE: multidevice_schedule and schedule are not defined in the FusionDefinition base class. 

from typing import Callable
has_multidevice_schedule_fn = hasattr(self, 'multidevice_schedule') and isinstance(self.multidevice_schedule, Callable)
has_schedule_fn = hasattr(self, 'schedule') and isinstance(self.schedule, Callable)

# multidevice_schedule is incompatible with schedule, so check that only one is defined.
assert has_multidevice_schedule_fn ^ has_schedule_fn

if has_multidevice_schedule_fn:
  self.multidevice_schedule()

if has_schedule_fn:
   self._setup_schedule(inputs)
   self.schedule()
   self._finalize_schedule(inputs)
```

Comment (rdspring1) [reply nvfuser/__init__.py:216]: > schedule() should be implemented by child class! 

I've been using decorators to define the schedule function so this ^^^ child class requirement is cumbersome.

Decorator example:
```python
def schedule_decorator(fd):
  def schedule_fn():
    # do some scheduling
  fd.schedule = schedule_fn
  return fd
```

**Review by rdspring1 — COMMENTED** [review]
LGTM.

I gave a suggestion for a different API to avoid potential conflicts with the `schedule` function. `multidevice_schedule` doesn't conflict with the main path, so no strong objections as-is.

**Comment by wujingyue** [comment]
!build

File: nvfuser/__init__.py (hunk 1 above)
Comment (wujingyue) [reply nvfuser/__init__.py]: I believe it's not XOR but NAND. Anyhow, I think the latest commit has addressed your concern.

Comment (rdspring1) [reply nvfuser/__init__.py]: True. Both functions are optional.

**Review by rdspring1 — APPROVED** [review]

**Pull request merged**